crates/tracing-instrumentation/src/lib.rs

```diff
@@ -59,20 +62,129 @@ pub enum Error {
+
+    let scheme = uri
+        .scheme()
+        .ok_or_else(|| bad_endpoint(format!("{endpoint}: no scheme")))?
+        .to_string();
+
+    // Tokenize the scheme on '+' to determine the type of exporter.
+    let mut scheme_tokens: Vec<&str> = scheme.split('+').collect();
+    scheme_tokens.sort();
+
+    enum Transport {
+        Tonic, // gRPC
+        Http,  // HTTP(s)
+    }
+
+    // Map specific token combinations to ultimate endpoint scheme, exporter
+    // transport, and exporter protocol.
+    let (final_scheme, use_transport, use_protocol) = match scheme_tokens.as_slice() {
+        ["http"] => ("http", Transport::Tonic, Protocol::Grpc),
+        ["https"] => ("https", Transport::Tonic, Protocol::Grpc),
+        ["grpc"] => ("http", Transport::Tonic, Protocol::Grpc),
+        ["grpc", "otlp"] => ("http", Transport::Tonic, Protocol::Grpc),
+        ["http", "otlp"] => ("http", Transport::Http, Protocol::HttpBinary),
+        ["https", "otlp"] => ("https", Transport::Http, Protocol::HttpBinary),
+        ["http", "otlp", "proto"] => ("http", Transport::Http, Protocol::HttpBinary),
+        ["https", "otlp", "proto"] => ("https", Transport::Http, Protocol::HttpBinary),
+        ["http", "json", "otlp"] => ("http", Transport::Http, Protocol::HttpJson),
+        ["https", "json", "otlp"] => ("https", Transport::Http, Protocol::HttpJson),
+        _ => return Err(bad_endpoint(format!("{endpoint}: invalid scheme"))),
+    };
+
+    // Reconstruct the endpoint with the ultimate scheme from above.
+    let endpoint = http::Uri::builder()
+        .scheme(final_scheme)
+        .authority(uri.authority().map(|a| a.as_str()).unwrap_or(""))
+        .path_and_query(uri.path_and_query().map(|pq| pq.as_str()).unwrap_or("/"))
+        .build()
+        .map_err(|e| bad_endpoint(format!("rebuild endpoint: {e}")))?
+        .to_string();
+
+    // Build the exporter as specified.
+    let exporter = match use_transport {
+        Transport::Tonic => {
+            let metadata_headers: MetadataMap = MetadataMap::from_headers(HeaderMap::from_iter(
+                HashMap::from(tracing_headers.clone()),
+            ));
+            SpanExporter::builder()
+                .with_tonic()
+                .with_protocol(use_protocol)
+                .with_tls_config(ClientTlsConfig::new().with_native_roots()) // TODO: maybe only when https?
+                .with_metadata(metadata_headers)
+                .with_endpoint(&endpoint)
+                .build()
+                .map_err(|e| bad_endpoint(format!("build gRPC exporter: {e}")))?
+        }
+
+        Transport::Http => {
+            let client = reqwest::Client::new();
+            let string_headers: HashMap<String, String> = HashMap::from(tracing_headers.clone())
+                .iter()
+                .map(|(k, v)| (k.as_str().to_string(), v.to_str().unwrap_or("").to_string()))
+                .collect();
+            SpanExporter::builder()
+                .with_http()
```

Comment:
do we need to setup tls config here?

Reply:
Good catch. Potentially, yes. Will add.

Reply:
After review I think this should be fine as-is. In this case, with_http should delegate stuff like TLS to the client provided via with_http_client, and [reqwest::Client::new](https://docs.rs/reqwest/latest/reqwest/struct.Client.html#method.new) should pick up the appropriate TLS config automatically.

Reply:
I updated the with_http branch to build its HTTP client with TLS options that I believe should be the same as the with_tonic branch, i.e. drawing from rustls-tls-native-roots.